### PR TITLE
feature: define an implement hamt interface

### PR DESCRIPTION
**Pull request by frrist**

### CODE REVIEW

File: actors/builtin/multisig/multisig_actor.go [hunk 1]
```diff
@@ -2,7 +2,7 @@ package multisig
 
 import (
 	"encoding/binary"
-	"fmt"
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: import grouping

File: actors/util/adt/hamt.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+type Store interface {
+	Context() context.Context
+	hamt.CborIpldStore
+}
+
+type Keyer interface {
+	Key() string
+}
+
+type HamtStore struct {
```
Comment (anorth) [inline actors/util/adt/hamt.go]: I see the point of this being to abstract many of the HAMT implementation details. To that end, I think we should call this `Map`.

File: actors/builtin/multisig/multisig_actor.go [hunk 3]
```diff
@@ -89,13 +95,17 @@ func (a *MultiSigActor) Propose(rt vmr.Runtime, params *ProposeParams) *ProposeR
...
+	actorHamt := adt.NewHamt(vmr.AsStore(rt), st.PendingTxns)
+	if err := actorHamt.Put(txnID, &MultiSigTransaction{
 		To:       params.To,
 		Value:    params.Value,
 		Method:   params.Method,
 		Params:   params.Params,
 		Approved: []addr.Address{callerAddr},
-	})
+	}); err != nil {
+		rt.AbortStateMsg(err.Error())
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]: Please don't use AbortStateMsg, I'm trying to kill it. Use Abort().

We might need a new error code in common.go, ErrIllegalState

File: actors/builtin/multisig/multisig_actor.go [hunk 4]
```diff
@@ -89,13 +95,17 @@ func (a *MultiSigActor) Propose(rt vmr.Runtime, params *ProposeParams) *ProposeR
 	txnID := st.NextTxnID
 	st.NextTxnID += 1
 
-	st.PendingTxns = setPendingTxn(rt, st.PendingTxns, txnID, MultiSigTransaction{
+	actorHamt := adt.NewHamt(vmr.AsStore(rt), st.PendingTxns)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go]:
```suggestion
	txnMap := adt.NewMap(vmr.AsStore(rt), st.PendingTxns)
```

File: actors/util/adt/hamt.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
+package adt
+
+import (
+	"context"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-hamt-ipld"
+	typegen "github.com/whyrusleeping/cbor-gen"
+)
+
+type Store interface {
```
Comment (anorth) [inline actors/util/adt/hamt.go]: Document these interfaces.

File: actors/builtin/multisig/multisig_actor.go (hunk 3 above)
Comment (frrist) [reply actors/builtin/multisig/multisig_actor.go]: I've added the `ErrIllegalState` to common.go and instumented it for hamt operation failures in this file.

**Pull request merged**